File: src/engine/commands/gocommand.cpp
#include "gocommand.h"

#include "portal.h"


#define super Command

GoCommand::GoCommand(QObject *parent) :
    super(parent) {

    setDescription("Go to or enter an exit.\n"
                   "\n"
                   "Examples: go north, go to tower, enter door");
}

GoCommand::~GoCommand() {
}

void GoCommand::execute(Player *player, const QString &command) {

    super::prepareExecute(player, command);

    if (alias() == "go") {
        takeWord("to", IfNotLast);
    }

    ObjectDescription description = takeObjectsDescription();
    GameObjectPtrList portals;
    for (const GameObjectPtr &portalPtr : currentRoom()->portals()) {
        Portal *portal = portalPtr.cast<Portal *>();
        QString loweredName = portal->nameFromRoom(player->currentRoom()).toLower();
        for (const QString &word : loweredName.split(' ')) {
            if (word.startsWith(description.name)) {
                portals.append(portal);
                break;
            }
        }
    }
    if (description.position > 0) {
        if (description.position <= (int) portals.length()) {
            GameObjectPtr selected = portals[description.position - 1];
            portals.clear();
            portals.append(selected);
        } else {
            portals.clear();
        }
    }

    if (!requireSome(portals, "Go where?")) {
        return;
    }

    player->go(portals[0]);
}

File: data/commands/gocommand.js
function GoCommand() {
    Command.call(this);

    this.setDescription("Go to or enter an exit.\n" +
                        "\n" +
                        "Examples: go north, go to tower, enter door, go forward");
}

GoCommand.prototype = new Command();
GoCommand.prototype.constructor = GoCommand;

GoCommand.prototype.execute = function(character, command) {

    this.prepareExecute(character, command);

    if (this.alias === "go") {
        var direction = this.peekWord();
        if (direction === "forward" || direction === "back" ||
            direction === "left" || direction === "right") {
            character.execute(direction);
        }

        this.takeWord("to", Options.IfNotLast);
    }

    var portal = this.takeObject(character.currentRoom.portals);
    if (!this.requireSome(portal, "Go where?")) {
        return;
    }

    character.go(portal);
};

CommandRegistry.registerCommand("enter", new GoCommand());
CommandRegistry.registerCommand("go", new GoCommand());
